src/encode.cpp

```diff
@@ -458,6 +464,140 @@ static void encodeBlock(
+          verifComplexity = Complexity::TENSOR;
+        } else if (operand_ty.isa<mlir::MemRefType>()) {
+          verifComplexity = Complexity::MEMREF;
+        } else if (!operand_ty.isIntOrIndexOrFloat()) {
+          throw UnsupportedException("Invalid operand type");
+        }
+      }
+
+      switch (verifComplexity) {
+        case Complexity::SCALAR: {
+          vector<Sort> smtDomain;
+          smtDomain.reserve(domain.size());
+          transform(domain.cbegin(), domain.cend(),
+            back_inserter(smtDomain), getScalarSort);
+
+          if (range.isa<mlir::TensorType>()) {
+            throw UnsupportedException(
+            "Function that returns tensor is not supported yet");
+          }
+          if (range.isa<mlir::MemRefType>()) {
+            throw UnsupportedException(
+            "Function that returns memref is not supported yet");
+          }
+
+          const auto smtRange = getScalarSort(range);
+          FnDecl decl(smtDomain, smtRange, string(name) + "_tvfn");
+
+          return DeclaredUF(move(domain), move(range), move(decl),
+                                  verifComplexity);
+        }
+        case Complexity::TENSOR: {
+          throw UnsupportedException(
+            "Function call with tensor operand(s) is unsupported");
+        }
+        case Complexity::MEMREF: {
+          throw UnsupportedException(
+            "Function call with memref operand(s) is unsupported");
+        }
+      }
+
+      llvm_unreachable("Invalid verification complexity");
+    }
+
+    ValueTy apply(const vector<ValueTy>& operands) const {
+      vector<Expr> operandExprs;
+      operandExprs.reserve(operands.size());
+
+      transform(operands.cbegin(), operands.cend(),
+        back_inserter(operandExprs),
+        [](const auto& operand) { return getExpr(operand); });
+      const auto fn_output = decl.apply(operandExprs);
+
+      if (range.isa<mlir::FloatType>()) {
+        return Float(fn_output, range);
+      } else if (range.isa<mlir::IntegerType>()) {
+        return Integer(fn_output);
+      } else if (range.isIndex()) {
+        return Index(fn_output);
+      } else {
+        throw UnsupportedException(
+          "Function that returns non-scalar value is not supported yet");
+      }
+    }
+  };
+
+  map<string, DeclaredUF, std::less<>> calleeMap;
+
+  optional<DeclaredUF> getFunction(const string_view name) {
+    auto fn_itr = calleeMap.find(name);
+    if (fn_itr != calleeMap.end()) {
+      return fn_itr->second;
+    } else {
+      return nullopt;
+    }
+  }
+
+  bool declareFunction(vector<mlir::Type>&& domain, mlir::Type&& range,
+                        const string_view name) {
+    if (getFunction(name)) {
+      // no-op if there already exists a function of the same name
+      return false;
+    } else {
+      calleeMap.insert({string(name),
+                        DeclaredUF::declare(move(domain), move(range), name)});
+      return true;
+    }
+  }
+}
```

Comment:
Let's split these into a separate header.

Reply:
Moved to `function.h` and `function.cpp`